docker/importer/importer.py

```diff
@@ -307,50 +315,82 @@ def _process_updates_git(self, source_repo):
     source_repo.last_synced_hash = str(repo.head.target)
     source_repo.put()
 
-  def _process_updates_bucket(self, source_repo):
+    logging.info("Finish processing git: %s", source_repo.name)
+
+  def _process_updates_bucket(self, source_repo: osv.SourceRepository):
     """Process updates from bucket."""
     # TODO(ochang): Use Pub/Sub change notifications for more efficient
     # processing.
+    logging.info("Begin processing bucket: %s", source_repo.name)
 
     ignore_last_import_time = source_repo.ignore_last_import_time
     if ignore_last_import_time:
       source_repo.ignore_last_import_time = False
       source_repo.put()
 
+    # First retrieve a list of files to parallel download
     storage_client = storage.Client()
-    for blob in storage_client.list_blobs(source_repo.bucket):
-      if not _is_vulnerability_file(source_repo, blob.name):
-        continue
-
-      logging.info('Bucket entry triggered for for %s/%s', source_repo.bucket,
-                   blob.name)
-      blob_bytes = blob.download_as_bytes()
-      try:
-        vulnerabilities = osv.parse_vulnerabilities_from_data(
-            blob_bytes,
-            os.path.splitext(blob.name)[1])
-      except Exception as e:
-        logging.error('Failed to parse vulnerability %s: %s', blob.name, e)
-        continue
-
-      def unchanged_vuln_exist(vuln):
-        bug = osv.Bug.get_by_id(vuln.id)
-        # Both times are not timezone aware
-        return bug and bug.import_last_modified == vuln.modified.ToDatetime()
-
-      if not ignore_last_import_time and all(
-          unchanged_vuln_exist(vuln) for vuln in vulnerabilities):
-        logging.info('Skipping updates for %s as modified date unchanged.',
-                     blob.name)
-        continue
-
-      original_sha256 = osv.sha256_bytes(blob_bytes)
-      self._request_analysis_external(source_repo, original_sha256, blob.name)
+    listed_blob_names = [
+        blob.name for blob in (storage_client.list_blobs(source_repo.bucket))
+    ]
+    batched_blob_names = [
```

Comment:
How much does batching get us if we're doing multithreading anyway? Can we just increase the multithreading count to get the same perf? 

Also I'm pretty sure a `.IN` query in ndb just runs parallel queries under the hood, so we're not actually saving on # of requests if that's what we were after. 

If it doesn't buy us much, I reckon let's just do the multithreading and keep the code a lot simpler.

Reply:
Did a test, for about 50 entries, it takes roughly double the time of .IN, but it's still a lot faster than running one by one.

Reply:
Weird, there must be some additional optimization happening under the hood for .IN. Let's start with this since it's already a big improvement.

Reply:
Now using the multithreading code instead of .IN, it does simplify the code quite a bit, so it's probably worth it in terms of complexity to performance tradeoff.